nni/retiarii/execution/cgo_engine.py

```diff
@@ -0,0 +1,151 @@
+from ..integration import send_trial, receive_trial_parameters, get_advisor
+from .logical_optimizer.logical_plan import LogicalPlan, PhysicalDevice
+from .logical_optimizer.opt_dedup_input import DedupInputOptimizer
+
+from .base import BaseGraphData
+
+_logger = logging.getLogger('nni.msg_dispatcher_base')
+class CGOExecutionEngine(AbstractExecutionEngine):
+    def __init__(self, n_model_per_graph = 4) -> None:
+        self._listeners: List[AbstractGraphListener] = []
+        self._running_models: Dict[int, Model] = dict()
+        self.logical_plan_counter = 0
+        self.n_model_per_graph = n_model_per_graph
+        self._optimizers = [DedupInputOptimizer()]
+        self._original_models = {}
+        self._original_model_to_multi_model = {}
+
+        # register advisor callbacks
+        advisor = get_advisor()
+        advisor.send_trial_callback = self._send_trial_callback
+        advisor.request_trial_jobs_callback = self._request_trial_jobs_callback
+        advisor.trial_end_callback = self._trial_end_callback
+        advisor.intermediate_metric_callback = self._intermediate_metric_callback
+        advisor.final_metric_callback = self._final_metric_callback
+
+    
+    def add_optimizer(self, opt):
+        self._optimizers.append(opt)
+
+    def submit_models(self, *models: List[Model]) -> None:
+        _logger.info(f'{len(models)} Models are submitted')
+        logical = self._build_logical(models)
+        
+        for opt in self._optimizers:
+            opt.convert(logical)
+        
+        phy_models_and_placements = self._assemble(logical)
+        for model, placement, grouped_models in phy_models_and_placements:
+            data = BaseGraphData(codegen.model_to_pytorch_script(model, placement=placement),
+                            model.training_config.module, model.training_config.kwargs)
+            for m in grouped_models:
+                self._original_models[m.model_id] = m
+                self._original_model_to_multi_model[m.model_id] = model
+            self._running_models[send_trial(data.dump())] = model
+            
+        # for model in models:
+        #     data = BaseGraphData(codegen.model_to_pytorch_script(model),
+        #                          model.config['trainer_module'], model.config['trainer_kwargs'])
+        #     self._running_models[send_trial(data.dump())] = model
+    
+    def _assemble(self, logical_plan : LogicalPlan) -> List[Tuple[Model, PhysicalDevice]]:
+        # unique_models = set()
+        # for node in logical_plan.graph.nodes:
+        #     if node.graph.model not in unique_models:
+        #         unique_models.add(node.graph.model)
+        # return [m for m in unique_models]
+        grouped_models : List[Dict[Model, PhysicalDevice]] = AssemblePolicy().group(logical_plan)
+        phy_models_and_placements = []
+        for multi_model in grouped_models:
+            model, model_placement = logical_plan.assemble(multi_model)
```

Comment:
it is not very clear that what `AssemblePolicy()` should do and what `logical_plan.assemble` should do

Reply:
`AssemblePolicy` decides which models are grouped in one physical model, i.e., each `multi_model`.

`logical_plan.assemble(multi_model)` extracts the models in `multi_model` from the `logical_plan`, and replace all logical nodes to physical nodes associated with the models in `multi_model`. nodes not in `multi_model` will not exist in the physical model after `logical_plan.assemble`.

For example, a logical `DedupInputNode` contains all models from `submit_models()` in the `logical_plan`, `logical_plan.assemble(multi_model)` will generate one input node connected to successor nodes of models in `multi_model`.

Reply:
got it, thanks